### DISCUSSION TITLE
x-mask plugin: get unmasked value

### DISCUSSION THREAD

**Discussion by manhthang2504** [alpinejs/alpine#2867] [1. Help]
I am using the [Mask plugin](https://alpinejs.dev/plugins/mask) for formatting money value in a input field which is also a Livewire nested model.
```
@foreach($classes as $index => $class)
<input x-mask:dynamic="$money($input)" wire:model="classes.{{ $index }}.paying_amount">
@endif
```
This model have validation rule:
```
protected $rules = [
        
        'classes.*.paying_amount' => 'numeric',
      
    ];
```
And it always raise validation error because the formated amount (say: 1,000,000) is not numeric but text.

Is there any way to bind the unmasked value (i.e: 1000000, instead of 1,000,000), or at least how to get the unmasked value using Alpine and/or Mask plugin?

**Comment by tomgobich** [comment]
Came across the same issue using Adonis, and searching for a solution brought me to your question. Thought I'd drop my solution in case anyone else follows the same path as me.

```
<div x-data="{ 
  value: {{ value }} ?? '',
  displayValue: {{ value }} ?? '',
  demaskValue(value) {
    this.value = value.replaceAll('$', '').replaceAll(',', '')
  }
}" class="-mr-2">
  <input 
    type="text" 
    x-mask:dynamic="'$' + $money($input)"
    :value="displayValue"
    @input="demaskValue($el.value)"
    placeholder="$0.00"
  />
  <input 
    type="hidden" 
    name="{{ name }}"
    x-model="value"
  />
  
  {{-- for testing purposes --}}
  <span x-text="value"></span>
</div>
```

Essentially if you use two inputs, one for the unmasked value and one for the displayed value, you can remove all masked characters on input and use the hidden unmasked value to post out with your form submission.

Here's the result with the unmasked value displayed in the span and the masked value displayed in the text input
![Screen Shot 2022-09-04 at 11 51 35 AM](https://user-images.githubusercontent.com/19398726/188322256-2f065ca8-51ab-4bb3-9b89-6a44c19d9497.png)

**Reply by AndreaB93** [reply]
Hello, i would use your solution because i've the same issue with mask plugin of alpine but i've get this error: 

![image](https://github.com/alpinejs/alpine/assets/30800117/9fc2f870-9bfe-483b-aeae-d975a9302b97)

Could you help me about it? That is my first time using alpine.

Thanks a lot!

**Reply by ekwoka** [reply]
Your templating isn't being swapped out. That makes it not an Alpine issue. But whatever your templating is.

**Reply by ezehkingsleyuchenna** [reply]
> <div x-data="{ 
>   value: {{ value }} ?? '',
>   displayValue: {{ value }} ?? '',
>   demaskValue(value) {
>     this.value = value.replaceAll('$', '').replaceAll(',', '')
>   }
> }" class="-mr-2">
>   <input 
>     type="text" 
>     x-mask:dynamic="'$' + $money($input)"
>     :value="displayValue"
>     @input="demaskValue($el.value)"
>     placeholder="$0.00"
>   />
>   <input 
>     type="hidden" 
>     name="{{ name }}"
>     x-model="value"
>   />
>   
>   {{-- for testing purposes --}}
>   <span x-text="value"></span>
> </div>

Try adding an initial value to the displayValue.
LIVEWIRE property `amount`
~~~
<div x-data="{ 
  value: $wire.entangle('amount'),
  displayValue: '',
  demaskValue(value) {
    this.value = value.replaceAll('$', '').replaceAll(',', '')
  }
}" class="-mr-2" x-init="displayValue = value">
  <input 
    type="text" 
    x-mask:dynamic="'$' + $money($input)"
    :value="displayValue"
    @input="demaskValue($el.value)"
    placeholder="$0.00"
  />
  
  {{-- for testing purposes --}}
  <span x-text="value"></span>
</div>

**Comment by rustomenglish** [comment]
On the PHP side, using `prepareForValidation` in either the Livewire Form or Component works for me:

```php
    protected function prepareForValidation($attributes)
    {
        // Unmask values
        $attributes['amount'] = (float) str_replace(',', '', $this->amount);

        return $attributes;
    }
```
Then in a save function:

```php
    public function save()
    {
        $data = $this->validate();
        //...
```

**Reply by poldixd** [reply]
But then it changes the value that is sent to the frontend again, doesn't it?

**Comment by ekwoka** [comment]
I believe a current PR would expose it as an export from the mask plugin package where you could pass the input and the mask and it would give back the result with full mask removed.

Of note, for masks, like money that would do `1,000.49` it would give back `100049` which might be good or not...

**Comment by cwyrwas** [comment]
Currently dealing with this right now. I can't seem to find much information, but I came across this discussion. Was a native solution ever implemented for this? I'm trying to use x-model on a currency input, but I need the unmasked float value for calculations.

**Reply by ekwoka** [reply]
You can just use a getter that gives you the number.

**Reply by cwyrwas** [reply]
That makes sense. I ultimately ended up just going that route. Thanks again!
